tests/std/tests/P2136R3_invoke_r/test.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <functional>
+#include <type_traits>
+
+using namespace std;
+
+// TRANSITION, DevCom-1457457
+namespace detail {
+    static constexpr bool permissive() {
+        return false;
+    }
+
+    template <class>
+    struct DependentBase {
+        static constexpr bool permissive() {
+            return true;
+        }
+    };
+
+    template <class T>
+    struct Derived : DependentBase<T> {
+        static constexpr bool test() {
+            return permissive();
+        }
+    };
+} // namespace detail
+constexpr bool is_permissive = detail::Derived<int>::test();
+
+constexpr int square(int n) {
+    return n * n;
+}
+constexpr int square_noexcept(int n) noexcept {
+    return n * n;
+}
+
+struct Thing {
+    int n = 0;
+    constexpr int& foo() {
+        return n;
+    }
+};
+
+constexpr bool is_rvalue(const int&) {
+    return false;
+}
+constexpr bool is_rvalue(int&&) {
+    return true;
+}
+
+constexpr bool test_invoke_r() {
+    auto v1 = invoke_r<long int>(square, 3);
+    assert(v1 == 9L);
+    static_assert(is_same_v<decltype(v1), long int>);
+
+    auto v2 = invoke_r<double>([]() -> int { return 5; });
+    assert(v2 == 5);
+    static_assert(is_same_v<decltype(v2), double>);
+    static_assert(is_void_v<decltype(invoke_r<void>(square, 1))>);
+
+    // TRANSITION, DevCom-1457457
+    static_assert(noexcept(invoke_r<int>(square, 3)) == is_permissive, "invoke_r<int>(square, 3) is noexcept");
+    static_assert(noexcept(invoke(square, 3)) == is_permissive, "invoke(square, 3) is noexcept");
+#ifdef __cpp_noexcept_function_type
+    static_assert(noexcept(invoke_r<int>(square_noexcept, 3)), "invoke_r<int>(square_noexcept, 3) isn't noexcept");
+    static_assert(noexcept(invoke(square_noexcept, 3)), "invoke(square_noexcept, 3) isn't noexcept");
+#endif
+
+    Thing thing;
+    invoke_r<void>(&Thing::n, thing); // no nodiscard warning
+    assert(!is_rvalue(invoke(&Thing::foo, thing)));
+    assert(is_rvalue(invoke_r<int>(&Thing::foo, thing)));
+
+    return true;
+}

```

Comment:
@cdacamar suggests adding test coverage for a couple more cases (which are extensively exercised by the underlying `invoke` machinery, we just need to verify that everything is being perfectly forwarded down to it):
1. Calling `invoke_r` on a function object that updates its state should update the original object (i.e. no copies are made and then dropped on the floor). Would be as simple as a lambda with an init-capture that increments a counter.
2. The callable object and its arguments should be perfectly forwarded. Could verify with a function object that has ref-qualified `operator()` overloads to sense the object's value category (and you could also make it templated on an argument type, to verify that value category).

Reply:
Not sure I fully understand `#2`.

Reply:
I was thinking of `template <typename T> operator()(T&&) ref-qualifier` where we can inspect `T` to see whether it was called with an lvalue or an rvalue.

Reply:
Forgive me if this is a dumb question but isn't that more verbose than what I have? (assuming you mean with an if constexpr on value category?) Or is what I have not correct?

Reply:
I'll take a look, thanks!

Reply:
Your ref-qualification inspection code is great, thanks - indeed less verbose than what I was thinking of.

I've pushed a small commit for (1) - a lambda with a ref-capture doesn't contain modifiable state within itself (the modifiable state is external, so copying the lambda won't have harmful effects). I've changed this to be an init-capture, then renamed the `count` to `lvalue` since its purpose has changed.